### PR TITLE
extending tools section

### PR DESCRIPTION
**Pull request by choldgraf**
This is a long synthesis of a meeting that @yuvipanda, @aculich, Jamie W, and I had about a week ago. The goal was to give a general high-level view of all the components of the jupyterhub pipeline. This PR was synthesized from these notes:

https://paper.dropbox.com/doc/Jupyterhub-High-level-setup-6Oe1ubmCBy9XORWJ2PnbH

There are likely still some holes in there, but I think this is a good start at getting people oriented to what these tools are and what they're useful for. It replicates some information that was in the glossary, so I removed those items, but if we like maybe it's better to have them link to this page where appropriate. I also think this can be extended further by having something like a workflow diagram, but that seems like too much for one PR :)

### CODE REVIEW

**Comment by choldgraf** [comment]
@yuvipanda any thoughts on whether this looks correct? or do you just want me to merge and we can spot-check it later?

File: doc/source/getting-started.rst [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+The goal of JupyterHub is to make it easy and straightforward to
+create custom computing environments that can be accessed remotely (e.g.,
+at a specific URL) by multiple users. Many of these pieces can be a little
+confusing, and this website is meant as an assistant to guide you through
+the process of setting up your JupyterHub.
+
+These materials are meant to help you connect the following things:
+
+* A cloud provider (Google Cloud, Azure, SC3)
```
Comment (yuvipanda) [inline doc/source/getting-started.rst]: what is SC3?

File: doc/source/getting-started.rst [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+* Kubernetes (to manage resources on the cloud)
+* Helm (to configure and control kubernetes)
+* Docker (to use containers that standardize computing environments)
+* JupyterHub (to manager user accounts and deploy Docker images)
```
Comment (yuvipanda) [inline doc/source/getting-started.rst]: I think (to manage users and deploy jupyter notebooks) is a better fit.

File: doc/source/tools.rst [hunk 3]
```diff
@@ -1,30 +1,259 @@
...
+tools in greater detail in order to provide some more contextual
+information.
 
-- Google Cloud will provide the computing power that we’ll use. This is a service provided by Google that essentially allows us to use some of their computers. Fortunately, their computers are fancy, with lots of tools for scaling our usage up or down depending on our needs. There are lots of other cloud services out there (such as Microsoft Azure and Amazon EC2) but we’ll focus on Google’s for now.
-  - You can access Google Cloud from its web console or via the “gcloud” SDK that you can download to your computer.
+Cloud Computing Providers
+-------------------------
 
-`Kubernetes <https://kubernetes.io/>`_
+These are the organizations that host the computers on which your
```
Comment (yuvipanda) [inline doc/source/tools.rst]: This seems confusing and kinda strange to me...

File: doc/source/tools.rst [hunk 4]
```diff
@@ -1,30 +1,259 @@
...
 
-- Docker is a technology for “containerized” computing environments. This basically means creating a very specific combination of hardware + software that can be easily moved to any computer. It’s useful for standardizing the environment in which development happens. We use this to generate live computing environments that users of a JupyterHub will experience. They’ll all have the same basic set of files + packages.
+Container Technology
+--------------------
 
-`Helm <https://github.com/kubernetes/helm/blob/master/docs/charts.md>`_
+Container technology is essentially the idea of bundling all of the
+necessary components to run a piece of software. There are many ways
+to do this, but one that we'll focus on is called Docker. Here are
+the main components of Docker:
```
Comment (yuvipanda) [inline doc/source/tools.rst]: I'd call these concepts rather than components.

File: doc/source/tools.rst [hunk 5]
```diff
@@ -1,30 +1,259 @@
...
+these resources are always available. If something goes down, kubernetes
+will try to automatically bring it back up.
+ 
+Kubernetes can only manage the computing resources that it is
+given. This means that it generally can **not** create new resources on its
+own (with the exception of disk space).
+
+There are three main types of objects in Kubernetes:
+
+Processes
```
Comment (yuvipanda) [inline doc/source/tools.rst]: This isn't a kubernetes concept / object. Not sure where it oculd go tho

File: doc/source/tools.rst [hunk 6]
```diff
@@ -1,30 +1,259 @@
...
+per machine, how much RAM).
+Then, it handles the resources on the cluster and ensures that
+these resources are always available. If something goes down, kubernetes
+will try to automatically bring it back up.
+ 
+Kubernetes can only manage the computing resources that it is
+given. This means that it generally can **not** create new resources on its
+own (with the exception of disk space).
+
+There are three main types of objects in Kubernetes:
```
Comment (yuvipanda) [inline doc/source/tools.rst]: I wouldn't call these main - I would say these are the things we care about most.

Comment (yuvipanda) [inline doc/source/tools.rst]: Also you have way more than 3 here :D https://www.youtube.com/watch?v=vt0Y39eMvpI

File: doc/source/tools.rst [hunk 7]
```diff
@@ -1,30 +1,259 @@
...
+service that you want to run, which makes things easier to manage and
+keeps things more stabl.
+
+Deployments
+***********
+
+A deployment is a collection of pods on kubernetes. It is how kubernetes
+knows exactly what containers and what machines need to be running at all
+times. For example, if you have two pods: one that does the authenticating
+described above, and another that manages a database, you can specify both
```
Comment (yuvipanda) [inline doc/source/tools.rst]: nope, you wouldn't do that. you'd have one deployment for a database, and one for your web app.

File: doc/source/tools.rst [hunk 8]
```diff
@@ -1,30 +1,259 @@
...
+IP addresses will be different, etc. However you don't want to have to
+re-orient yourself every time this happens. A Kubernetes service keeps
+track of all these changes on the backend, and provides a single address
+to manage your deployment.
+
+Namespace
+*********
+
+Finally, a `namespace <https://kubernetes.io/docs/admin/namespaces/>`_
+defines a collection of deployments. This will define
```
Comment (yuvipanda) [inline doc/source/tools.rst]: collections of various objects (not just deployments)

File: doc/source/tools.rst [hunk 9]
```diff
@@ -1,30 +1,259 @@
...
+the boundaries of hardware that the deployments have to exist within. It
+is generally the most "high-level" of the groups we've discussed thus far.
+For example, a a namespace could be a single class running with JupyterHub.
+It defines all the hardware that is available to students. It also defines
```
Comment (yuvipanda) [inline doc/source/tools.rst]: nope, doesn't have anything to do with hardware

File: doc/source/tools.rst [hunk 10]
```diff
@@ -1,30 +1,259 @@
...
+the JupyterHub machinery that glues together all of the student containers,
+manages disk space, etc.
+
+Persistent Volume Chain
```
Comment (yuvipanda) [inline doc/source/tools.rst]: Claim, rather than Chain.

File: doc/source/tools.rst [hunk 11]
```diff
@@ -1,30 +1,259 @@
...
+***********************
+
+Persistent Volume Chains are a way to have persistent storage without
+being tied down to one specific computer or machine. Kubernetes is
+about that flexibility, and that means that we don't want to lock ourselves
+in to a particular operating system just because our files are already
+on it. Persistent Volume Chains help deal with this problem by knowing
+how to convert files between filesystem types.
```
Comment (yuvipanda) [inline doc/source/tools.rst]: More like 'disk types' (AWS's disk, Google's disk, etc). Filesystem types have a very different meaning.

File: doc/source/tools.rst [hunk 12]
```diff
@@ -1,30 +1,259 @@
...
+
+Charts
+******
+
+The way that Helm controls kubernetes is with templates of structured
+information that specify some computational requirements.
+These templates are called "charts", or "helm charts". They contain
+all of the necessary information for kubernetes to generate either:
+
+- a deployment object
```
Comment (yuvipanda) [inline doc/source/tools.rst]: we generate a bunch of other stuff too, and we generate many deployments and services as well.

File: doc/source/tools.rst [hunk 13]
```diff
@@ -1,30 +1,259 @@
...
+
+
+Releases
+********
+
+A release is basically a specific instantiation of a helmchart inserted
+into a particular namespace. If you'd like to upgrade your
+kubernetes deployment (say, by changing the amount of RAM that each
+user should get), then you can change the helm chart, then re-deploy
+it to your kubernetes cluster. This generates a new release.
```
Comment (yuvipanda) [inline doc/source/tools.rst]: doesn't generate a new release - just a new version of the same release.

File: doc/source/tools.rst [hunk 14]
```diff
@@ -1,30 +1,259 @@
...
+
+
+JupyterHub
+----------
+
+JupyterHub is a way of utilizing the components above in order to
+provide computational environments that users can access remotely.
+It exists as a kubernetes deployment, which is comprised of multiple
```
Comment (yuvipanda) [inline doc/source/tools.rst]: there are two deployments (proxy and hub) each with one pod

File: doc/source/tools.rst [hunk 15]
```diff
@@ -1,30 +1,259 @@
...
+
+There are three main types of objects in Kubernetes:
+
+Processes
+*********
+Are any program that is running on a machine. For example,
+a Jupyter Notebook creates several processes that handle the
+execution of code and the display in the browser.
+
+Pods
```
Comment (yuvipanda) [inline doc/source/tools.rst]: https://kubernetes.io/docs/concepts/workloads/pods/pod-overview/ is pods

File: doc/source/tools.rst [hunk 16]
```diff
@@ -1,30 +1,259 @@
...
+  specified in its container image, and would be connected to the
+  outside world.
+* One that receives information from the authentication container, and
+  does something fancy with it (maybe it runs a python process).
+
+This is useful because it lets you compatmentalize the components of the
+service that you want to run, which makes things easier to manage and
+keeps things more stabl.
+
+Deployments
```
Comment (yuvipanda) [inline doc/source/tools.rst]: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/ is deployments

File: doc/source/tools.rst [hunk 17]
```diff
@@ -1,30 +1,259 @@
...
+checking the current state of the pods, and then comparing this with the
+original specification of the deployment. If there are differences between
+the current state vs. the specification of the deployment, Kubernetes will
+attempt to make changes until the current state matches the specification.
+
+**Note**: Users don't generally "create" deployments directly, they are
+instead generated from a set of instructions that are sent to Kubernetes.
+We'll cover this in the section on "Helm".
+
+Service
```
Comment (yuvipanda) [inline doc/source/tools.rst]: https://kubernetes.io/docs/concepts/services-networking/service/ services

File: doc/source/tools.rst [hunk 18]
```diff
@@ -1,30 +1,259 @@
...
+
+A service is simply a stable way of referring to a deployment. Kubernetes
+is all about intelligently handling dynamic and quickly-changing
+computational environments. This means that the hardware may change,
+IP addresses will be different, etc. However you don't want to have to
+re-orient yourself every time this happens. A Kubernetes service keeps
+track of all these changes on the backend, and provides a single address
+to manage your deployment.
+
+Namespace
```
Comment (yuvipanda) [inline doc/source/tools.rst]: Namespaces are https://kubernetes.io/docs/tasks/administer-cluster/namespaces/

File: doc/source/tools.rst (hunk 10 above)
Comment (yuvipanda) [inline doc/source/tools.rst]: https://kubernetes.io/docs/concepts/storage/persistent-volumes/ has more info

File: doc/source/getting-started.rst (hunk 1 above)
Comment (choldgraf) [inline doc/source/getting-started.rst]: lol, it is me typing too quickly and trying to type EC2 :P

Comment (choldgraf) [inline doc/source/getting-started.rst]: (thumbs up == addressed)

File: doc/source/tools.rst (hunk 5 above)
Comment (choldgraf) [inline doc/source/tools.rst]: added a bit more explanation

File: doc/source/tools.rst (hunk 3 above)
Comment (choldgraf) [inline doc/source/tools.rst]: tried to explain a bit better

File: doc/source/tools.rst (hunk 7 above)
Comment (choldgraf) [inline doc/source/tools.rst]: what's a good use case then to explain the value of a deployment? I'm just trying to motivate the idea that deployments aren't always 1 pod per deployment

File: doc/source/tools.rst (hunk 9 above)
Comment (choldgraf) [inline doc/source/tools.rst]: gimme an example then! ;-)

File: doc/source/tools.rst (hunk 12 above)
Comment (choldgraf) [inline doc/source/tools.rst]: what's the most user-relevant stuff to list?

**Comment by yuvipanda** [comment]
@choldgraf can you rebase? :(

**Comment by yuvipanda** [comment]
Maybe don't yet, I should actually review it first I guess.

**Comment by choldgraf** [comment]
@yuvipanda I responded to round 1 of reviews, lemme know if there's anything else to improve upon.

**Comment by willingc** [comment]
Merging and I'll be iterating on this in a new PR. Thanks @choldgraf :sunny:

**Pull request merged**